# Lab01: Lam quen voi Pytorch

## Bai 1. Tao Tensor co ban

### Code

In [3]:
import torch
a = torch.tensor(
[[1, 2, 3],
[4, 5, 6]],
dtype=torch.int32
)
print(a)

tensor([[1, 2, 3],
        [4, 5, 6]], dtype=torch.int32)


### Nhan xet Tensor A
- Kich thuoc: 2 hang x 3 cot
- So chieu: 2 chieu
- Kieu du lieu: 32 bit

# Bai 2. Tao cac Tensor thuong dung

### Code

In [4]:
import torch
a = torch.rand(3, 4)
b = torch.randn(3, 4)
c = torch.zeros(2, 3)
d = torch.ones(2, 3)
e = torch.eye(4)
print(a)
print(b)
print(c)
print(d)
print(e)

tensor([[0.2795, 0.2030, 0.5425, 0.2395],
        [0.5470, 0.7967, 0.1492, 0.1321],
        [0.1093, 0.6618, 0.2831, 0.7257]])
tensor([[-0.2697, -1.1301,  1.5644,  0.3916],
        [-1.0969,  1.2042, -2.0638, -1.5446],
        [ 1.7662, -1.1136, -1.4237, -0.5753]])
tensor([[0., 0., 0.],
        [0., 0., 0.]])
tensor([[1., 1., 1.],
        [1., 1., 1.]])
tensor([[1., 0., 0., 0.],
        [0., 1., 0., 0.],
        [0., 0., 1., 0.],
        [0., 0., 0., 1.]])


### Nhận xét

- `torch.rand()` tạo các số thực ngẫu nhiên trong khoảng [0, 1) theo phân phối đều.
- `torch.randn()` tạo các số thực ngẫu nhiên theo phân phối chuẩn, có giá trị trung bình gần **0** và có thể xuất hiện cả số âm lẫn số dương.
- Vì vậy, điểm khác nhau chính là:
  + `torch.rand()` chỉ sinh giá trị từ 0 đến nhỏ hơn 1.
  + `torch.randn()` có thể sinh giá trị âm, dương hoặc lớn hơn 1.

### Tao them mot Tensor 5x5 chua toan gia tri 10

In [6]:
import torch

f = torch.full((5, 5), 10)

print(f)

tensor([[10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10]])


## Bai 3. Kiem tra thong tin Tensor

### Code

In [7]:
import torch
a = torch.zeros(2, 3, 4)

print("Tensor:")
print(a)
print("Shape:", a.shape)
print("Size:", a.size())
print("Number of dimensions:", a.ndim)
print("Data type:", a.dtype)

Tensor:
tensor([[[0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.]],

        [[0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.]]])
Shape: torch.Size([2, 3, 4])
Size: torch.Size([2, 3, 4])
Number of dimensions: 3
Data type: torch.float32


- `torch.Size([2, 3, 4])` nghia la co 2 khoi, moi khoi co 3 hang, moi hang co 4 phan tu.

## Bai 4. Slicing va Indexing

### Code

In [8]:
import torch
a = torch.randn(3, 4, 5)
print("Original tensor:")
print(a)
print("\nTensor a[1]:")
print(a[1])
print("\nTensor a[1:, 2:4]:")

Original tensor:
tensor([[[ 0.3344,  0.6122, -1.0599, -1.2056, -0.4558],
         [ 0.7732, -2.0972, -0.8180,  0.6125,  1.4181],
         [ 0.6095, -0.4234, -0.4837,  1.1780, -0.2452],
         [ 0.3702, -0.4580,  0.7491,  0.4518, -0.0602]],

        [[-0.6724,  0.2119, -0.3787,  1.0896,  0.1885],
         [-0.3577, -0.0528, -0.0905, -0.2023,  1.0234],
         [ 0.6922, -0.7928, -0.3715,  0.0288, -1.3500],
         [-0.2011,  1.3329,  0.8958,  0.3252,  0.7736]],

        [[-1.7976,  0.2724, -1.3191, -0.7413,  0.5420],
         [ 0.1498, -1.3037, -0.3767,  0.0436, -0.5303],
         [ 0.5048, -2.6094,  0.0131, -0.5709, -2.1902],
         [-0.2800, -0.5983,  0.2018,  1.1541,  1.3495]]])

Tensor a[1]:
tensor([[-0.6724,  0.2119, -0.3787,  1.0896,  0.1885],
        [-0.3577, -0.0528, -0.0905, -0.2023,  1.0234],
        [ 0.6922, -0.7928, -0.3715,  0.0288, -1.3500],
        [-0.2011,  1.3329,  0.8958,  0.3252,  0.7736]])

Tensor a[1:, 2:4]:


### Nhận xét

- `a[1]` có shape `torch.Size([4, 5])` vì đã chọn một phần tử cụ thể ở chiều thứ nhất.
- `a[1:, 2:4]` có shape `torch.Size([2, 2, 5])`.
- Hai phần tử cuối cùng trên chiều thứ ba được lấy bằng: a[:, :, -2:]

## Bai 5. Reshape va Flatten

### Code

In [10]:
import torch
a = torch.randn(3, 4)
print("Original:")
print(a)
print("\nFlatten:")
print(a.ravel())
print("\nReshape to 3 x 2 x 2:")
print(a.reshape(3, 2, 2))

Original:
tensor([[-0.4490,  1.9609, -0.4281, -1.2592],
        [-1.3039, -0.4990, -0.7539, -0.5695],
        [ 0.3846,  1.0520, -1.1698,  0.7832]])

Flatten:
tensor([-0.4490,  1.9609, -0.4281, -1.2592, -1.3039, -0.4990, -0.7539, -0.5695,
         0.3846,  1.0520, -1.1698,  0.7832])

Reshape to 3 x 2 x 2:
tensor([[[-0.4490,  1.9609],
         [-0.4281, -1.2592]],

        [[-1.3039, -0.4990],
         [-0.7539, -0.5695]],

        [[ 0.3846,  1.0520],
         [-1.1698,  0.7832]]])


### Nhan xet
- So phan tu ban dau la: 12 (vi Tensor co kich thuoc 3x4)
- So phan tu sau resharp: 12 (resharp thanh 3x2x2)
- So phan tu sau khi resharp thanh 2x6:

In [11]:
c = a.reshape(2, 6)

print("\nReshape to 2 x 6:")
print(c)


Reshape to 2 x 6:
tensor([[-0.4490,  1.9609, -0.4281, -1.2592, -1.3039, -0.4990],
        [-0.7539, -0.5695,  0.3846,  1.0520, -1.1698,  0.7832]])


## Bai 6. Phep toan va nhan ma tran

### Code

In [13]:
import torch
a = torch.randn(2, 3)
b = torch.randn(2, 3)
print("a:")
print(a)
print("b:")
print(b)
print("\na + b:")
print(a + b)
print("\na / b:")
print(a / b)
print("\na^2:")
print(a ** 2)
print("\nMatrix multiplication:")
print(a @ b.T)

a:
tensor([[ 0.5607, -0.6813, -0.3056],
        [-0.0934,  0.4520, -0.9961]])
b:
tensor([[ 0.1152,  0.2921, -2.5336],
        [ 0.8432, -0.9042, -1.5793]])

a + b:
tensor([[ 0.6759, -0.3892, -2.8392],
        [ 0.7498, -0.4522, -2.5754]])

a / b:
tensor([[ 4.8667, -2.3327,  0.1206],
        [-0.1108, -0.4999,  0.6307]])

a^2:
tensor([[0.3144, 0.4642, 0.0934],
        [0.0087, 0.2043, 0.9922]])

Matrix multiplication:
tensor([[0.6398, 1.5714],
        [2.6450, 1.0857]])


### Nhận xét

- `a * b` là phép nhân từng phần tử.
  Hai Tensor phải có cùng shape hoặc có thể broadcast với nhau.

- `a @ b.T` là phép nhân ma trận
  Với:
  - `a` có shape `torch.Size([2, 3])`
  - `b.T` có shape `torch.Size([3, 2])`

- Khác biệt chính:
  - `a * b`: nhân từng phần tử tương ứng.
  - `a @ b.T`: thực hiện phép nhân ma trận theo quy tắc hàng x cột.

## Bai 7. Cac ham thong ke tren Tensor

### Code

In [14]:
import torch
a = torch.randn(3, 4)
print(a)
print("\nMean:")
print(torch.mean(a, dim=0))
print("\nStandard deviation:")
print(torch.std(a, dim=0))
print("\nCumulative sum:")
print(torch.cumsum(a, dim=0))

tensor([[ 0.6609, -0.8192, -0.6747,  0.7880],
        [-0.8288,  0.7351, -0.7446,  1.7010],
        [ 0.3221,  0.3229,  0.6419, -0.0269]])

Mean:
tensor([ 0.0514,  0.0796, -0.2592,  0.8207])

Standard deviation:
tensor([0.7809, 0.8052, 0.7811, 0.8644])

Cumulative sum:
tensor([[ 0.6609, -0.8192, -0.6747,  0.7880],
        [-0.1679, -0.0841, -1.4194,  2.4890],
        [ 0.1542,  0.2389, -0.7775,  2.4620]])


- Thay dim=0 bang dim=1

In [15]:
import torch
a = torch.randn(3, 4)
print(a)
print("\nMean:")
print(torch.mean(a, dim=1))
print("\nStandard deviation:")
print(torch.std(a, dim=1))
print("\nCumulative sum:")
print(torch.cumsum(a, dim=1))

tensor([[-0.3046,  0.9411,  0.8705,  0.5582],
        [ 0.3906, -1.8928, -0.3792,  1.9244],
        [-0.4079, -0.8692,  0.7756,  1.0124]])

Mean:
tensor([0.5163, 0.0107, 0.1277])

Standard deviation:
tensor([0.5720, 1.5898, 0.9098])

Cumulative sum:
tensor([[-0.3046,  0.6365,  1.5070,  2.0652],
        [ 0.3906, -1.5022, -1.8814,  0.0430],
        [-0.4079, -1.2772, -0.5016,  0.5108]])


### Nhan xet su khac biet:
- `dim=0` tinh toan theo chieu doc (theo cot)
- `dim=1` tinh toan theo chieu ngang (theo hang)

## Bai 8. Autograd co ban

### Code

In [17]:
import torch
x = torch.tensor(
3.6,
requires_grad=True
)
y = x * x
y.backward()
print("x =", x)
print("y =", y)
print("x.grad =", x.grad)

x = tensor(3.6000, requires_grad=True)
y = tensor(12.9600, grad_fn=<MulBackward0>)
x.grad = tensor(7.2000)


### Tinh thu cong dao ham y=x^2 tai x = 3.6
- Dao ham y=x^2
    => y' = 2x
- Tai x = 3.6: 
- 2 * 3.6 = 7.2


### Doi ham thanh y = 3*x**2 + 2*x + 1

In [18]:
import torch

x = torch.tensor(3.6, requires_grad=True)

y = 3 * x**2 + 2 * x + 1

y.backward()

print("x =", x)
print("y =", y)
print("x.grad =", x.grad)

x = tensor(3.6000, requires_grad=True)
y = tensor(47.0800, grad_fn=<AddBackward0>)
x.grad = tensor(23.6000)


- Gradient = 23.6

## Bai 9. Gradient Descent tim he so da thuc

### Code

In [20]:
import numpy as np
import torch
polynomial = np.poly1d([1, 2, 3])
N = 20
X = np.random.randn(N, 1) * 5
Y = polynomial(X)
XX = np.hstack([
    X * X,
    X,
    np.ones_like(X)
])
w = torch.randn(3, 1, requires_grad=True)
x = torch.tensor(XX, dtype=torch.float32)
y = torch.tensor(Y, dtype=torch.float32)
optimizer = torch.optim.NAdam([w], lr=0.01)
print("Initial coefficients:")
print(w)
for epoch in range(1000):
    optimizer.zero_grad()
    y_pred = x @ w
    mse = torch.mean(torch.square(y - y_pred))
    mse.backward()
    optimizer.step()
print("Learned coefficients:")
print(w)

Initial coefficients:
tensor([[-1.2785],
        [ 0.4671],
        [-1.3196]], requires_grad=True)
Learned coefficients:
tensor([[1.0176],
        [1.9998],
        [1.3179]], requires_grad=True)


### Nhận xét
- Hệ số học được xấp xỉ [1.0176, 1.9998, 1.3179].
- So với hệ số thật [1, 2, 3], hai hệ số đầu khá gần, còn hệ số thứ ba vẫn lệch khá nhiều.

### Vai trò các lệnh
- optimizer.zero_grad(): xóa gradient cũ trước mỗi vòng lặp.
- mse.backward(): tính gradient của hàm mất mát MSE.
- optimizer.step(): cập nhật các hệ số w dựa trên gradient đã tính.

## Bai 10. Giai he phuong trinh bang Autograd

### Code

In [1]:
import random
import torch
A = torch.tensor(random.random(), requires_grad=True)
B = torch.tensor(random.random(), requires_grad=True)
C = torch.tensor(random.random(), requires_grad=True)
D = torch.tensor(random.random(), requires_grad=True)
EPOCHS = 2000
optimizer = torch.optim.NAdam([A, B, C, D], lr=0.01)
for _ in range(EPOCHS):
    y1 = A + B - 9
    y2 = C - D - 1
    y3 = A + C - 8
    y4 = B - D - 2
    sqerr = y1*y1 + y2*y2 + y3*y3 + y4*y4
    optimizer.zero_grad()
    sqerr.backward()
    optimizer.step()
print("A =", A)
print("B =", B)
print("C =", C)
print("D =", D)

A = tensor(4.2086, requires_grad=True)
B = tensor(4.7913, requires_grad=True)
C = tensor(3.7913, requires_grad=True)
D = tensor(2.7913, requires_grad=True)


### Nhận xét
Nghiệm thu được:
- A ≈ 4.2086
- B ≈ 4.7913
- C ≈ 3.7913
- D ≈ 2.7913

Kiểm tra:
- A + B ≈ 9
- C - D ≈ 1
- A + C ≈ 8
- B - D ≈ 2

=> Nghiệm tìm được thỏa mãn bốn phương trình.

Bài toán có thể có nhiều nghiệm vì các phương trình không độc lập với nhau. Một số phương trình có thể suy ra từ các phương trình còn lại, nên hệ không xác định duy nhất một bộ giá trị A, B, C, D

## Bai 11. Viet chuong trinh PyTorch de thuc hien cac yeu cau sau:
- Tạo Tensor X có 24 phần tử và reshape lần lượt thành 4 × 6, 2 × 12 và 2 × 3 × 4.
- Tạo hai Tensor A, B kích thước 3 × 4; tính A+B, A*B, A/B, trung bình theo cột và độ lệch chuẩn theo cột.
- Cho y = 2x2 + 5x + 3. Dùng autograd tính dy/dx tại x = 2 và kiểm tra bằng phép tính thủ công.
- Thay đa thức trong Bài 9 bằng y = 2x2 - 3x + 5; huấn luyện và cho biết các hệ số tìm được.

### Yeu cau 1: Tạo Tensor X có 24 phần tử và reshape lần lượt thành 4 × 6, 2 × 12 và 2 × 3 × 4.

In [8]:
import torch
import numpy as np

In [9]:
X = torch.arange(24)

print("X ban dau:")
print(X)

print("\nReshape 4 x 6:")
print(X.reshape(4, 6))

print("\nReshape 2 x 12:")
print(X.reshape(2, 12))

print("\nReshape 2 x 3 x 4:")
print(X.reshape(2, 3, 4))

X ban dau:
tensor([ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11, 12, 13, 14, 15, 16, 17,
        18, 19, 20, 21, 22, 23])

Reshape 4 x 6:
tensor([[ 0,  1,  2,  3,  4,  5],
        [ 6,  7,  8,  9, 10, 11],
        [12, 13, 14, 15, 16, 17],
        [18, 19, 20, 21, 22, 23]])

Reshape 2 x 12:
tensor([[ 0,  1,  2,  3,  4,  5,  6,  7,  8,  9, 10, 11],
        [12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22, 23]])

Reshape 2 x 3 x 4:
tensor([[[ 0,  1,  2,  3],
         [ 4,  5,  6,  7],
         [ 8,  9, 10, 11]],

        [[12, 13, 14, 15],
         [16, 17, 18, 19],
         [20, 21, 22, 23]]])


### Yeu cau 2: Tạo hai Tensor A, B kích thước 3 × 4; tính A+B, A*B, A/B, trung bình theo cột và độ lệch chuẩn theo cột.

In [10]:
A = torch.randn(3, 4)
B = torch.randn(3, 4)

print("A:")
print(A)

print("\nB:")
print(B)

print("\nA + B:")
print(A + B)

print("\nA * B:")
print(A * B)

print("\nA / B:")
print(A / B)

print("\nTrung binh theo cot cua A:")
print(torch.mean(A, dim=0))

print("\nDo lech chuan theo cot cua A:")
print(torch.std(A, dim=0))

A:
tensor([[ 0.3194, -1.0361, -0.7518, -1.4616],
        [ 0.6158,  0.7944,  0.3937, -0.5853],
        [-1.1309, -1.0061, -0.4690, -0.0163]])

B:
tensor([[ 0.3229,  1.5077, -0.6506, -0.7832],
        [-0.5034,  0.2866, -0.7363, -0.2999],
        [ 0.2078,  0.5865, -0.0176,  0.1696]])

A + B:
tensor([[ 0.6424,  0.4715, -1.4024, -2.2448],
        [ 0.1124,  1.0810, -0.3426, -0.8851],
        [-0.9231, -0.4196, -0.4866,  0.1533]])

A * B:
tensor([[ 0.1032, -1.5621,  0.4891,  1.1447],
        [-0.3100,  0.2277, -0.2899,  0.1755],
        [-0.2350, -0.5901,  0.0083, -0.0028]])

A / B:
tensor([[ 0.9892, -0.6872,  1.1555,  1.8661],
        [-1.2232,  2.7722, -0.5347,  1.9518],
        [-5.4418, -1.7155, 26.6230, -0.0962]])

Trung binh theo cot cua A:
tensor([-0.0652, -0.4159, -0.2757, -0.6877])

Do lech chuan theo cot cua A:
tensor([0.9347, 1.0483, 0.5967, 0.7280])


In [11]:
print("\nTrung binh theo cot cua B:")
print(torch.mean(B, dim=0))

print("\nDo lech chuan theo cot cua B:")
print(torch.std(B, dim=0))


Trung binh theo cot cua B:
tensor([ 0.0091,  0.7936, -0.4682, -0.3045])

Do lech chuan theo cot cua B:
tensor([0.4476, 0.6363, 0.3926, 0.4764])


### Yeu cau 3: Cho y = 2x2 + 5x + 3. Dùng autograd tính dy/dx tại x = 2 và kiểm tra bằng phép tính thủ công.

Code:

In [12]:
x = torch.tensor(2.0, requires_grad=True)

y = 2 * x**2 + 5 * x + 3

y.backward()

print("y =", y)
print("Gradient tại x = 2:", x.grad)

y = tensor(21., grad_fn=<AddBackward0>)
Gradient tại x = 2: tensor(13.)


Tinh thu cong:


Ta co:  dy/dx = 4x + 5

Tai x = 2: 4 × 2 + 5 = 13

=> x.grad khoang 13

### Yeu cau 4: Thay đa thức trong Bài 9 bằng y = 2x2 - 3x + 5; huấn luyện và cho biết các hệ số tìm được.

In [13]:
import numpy as np
import torch

polynomial = np.poly1d([2, -3, 5])

N = 20
X = np.random.randn(N, 1) * 5
Y = polynomial(X)

XX = np.hstack([
    X * X,
    X,
    np.ones_like(X)
])

w = torch.randn(3, 1, requires_grad=True)

x = torch.tensor(XX, dtype=torch.float32)
y = torch.tensor(Y, dtype=torch.float32)

optimizer = torch.optim.NAdam([w], lr=0.01)

print("Initial coefficients:")
print(w)

for epoch in range(2000):
    optimizer.zero_grad()

    y_pred = x @ w

    mse = torch.mean((y - y_pred) ** 2)

    mse.backward()

    optimizer.step()

print("\nLearned coefficients:")
print(w)

Initial coefficients:
tensor([[ 0.9686],
        [-0.7509],
        [ 0.1368]], requires_grad=True)

Learned coefficients:
tensor([[ 2.0144],
        [-2.9693],
        [ 4.6619]], requires_grad=True)


- He so tim duoc: [2, -3, 5]

### End